File: android/src/main/java/com/onegini/mobile/sdk/reactnative/handlers/pins/PinAuthenticationEventEmitter.kt
package com.onegini.mobile.sdk.reactnative.handlers.pins

import com.facebook.react.bridge.Arguments
import com.facebook.react.modules.core.DeviceEventManagerModule
import com.onegini.mobile.sdk.reactnative.Constants
import com.onegini.mobile.sdk.reactnative.Constants.PinFlow
import com.onegini.mobile.sdk.reactnative.exception.OneginiWrapperError
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class PinAuthenticationEventEmitter @Inject constructor(private val deviceEventEmitter: DeviceEventManagerModule.RCTDeviceEventEmitter) {

  fun onPinOpen(profileId: String) {
    Arguments.createMap().apply {
      putString("action", Constants.PIN_NOTIFICATION_OPEN_VIEW)
      putString("flow", PinFlow.Authentication.toString())
      putString("profileId", profileId);
    }.also { dataMap ->
      deviceEventEmitter.emit(Constants.ONEWELCOME_PIN_AUTHENTICATION_NOTIFICATION, dataMap)
    }
  }

  fun onPinClose() {
    Arguments.createMap().apply {
      putString("action", Constants.PIN_NOTIFICATION_CLOSE_VIEW)
      putString("flow", PinFlow.Authentication.toString())
    }.also { dataMap ->
    deviceEventEmitter.emit(Constants.ONEWELCOME_PIN_AUTHENTICATION_NOTIFICATION, dataMap)
    }
  }

  fun onIncorrectPin(remainingAttempts: Int) {
    Arguments.createMap().apply {
      putString("action", Constants.PIN_NOTIFICATION_INCORRECT_PIN)
      putString("flow", PinFlow.Authentication.toString())
      putString("remainingFailureCount", remainingAttempts.toString())
      putInt("errorCode", OneginiWrapperError.WRONG_PIN_ERROR.code)
      putString("errorMsg", OneginiWrapperError.WRONG_PIN_ERROR.message)
    }.also { dataMap ->
      deviceEventEmitter.emit(Constants.ONEWELCOME_PIN_AUTHENTICATION_NOTIFICATION, dataMap)
    }
  }
}

File: ios/Handlers/EventEmitters/PinAuthenticationEventEmitter.swift
class PinAuthenticationEventEmitter {
    func onPinOpen(profileId: String) {
        let data = [
            "flow": PinFlow.authentication.rawValue,
            "action": PinAuthNotification.open.rawValue,
            "profileId": profileId
        ]
        BridgeConnector.shared?.sendBridgeEvent(eventName: .pinAuthNotification, data: data)
    }

    func onPinClose() {
        let data = [
            "flow": PinFlow.authentication.rawValue,
            "action": PinAuthNotification.close.rawValue
        ]
        BridgeConnector.shared?.sendBridgeEvent(eventName: .pinAuthNotification, data: data)
    }

    func onIncorrectPin(error: Error, remainingFailureCount: UInt) {
        let data = [
            "flow": PinFlow.authentication.rawValue,
            "action": PinAuthNotification.incorrectPin.rawValue,
            "errorType": error.code,
            "errorMsg": error.localizedDescription,
            "remainingFailureCount": remainingFailureCount
        ] as [String: Any]
        BridgeConnector.shared?.sendBridgeEvent(eventName: .pinAuthNotification, data: data)
    }
}
